File: game.of.life.v3/game.of.life.v3/RectangularInfinite2DGrid.cs
namespace game.of.life.v3
{
    using System.Linq;
    using System.Collections.Generic;

    public class RectangularInfinite2DGrid : IGrid
    {
        private readonly List<Cell> _cells = new List<Cell>();

        public RectangularInfinite2DGrid(params Cell[] cells)
        {
            _cells.AddRange(cells);
        }

        public IEnumerable<Cell> Cells { get { return _cells; } }

        public void Discover()
        {
            var newCells =
                (from cell in _cells
                from neighbour in GetNeighbours(cell)
                where !_cells.Contains(neighbour)
                 select neighbour).Distinct().ToArray();

            _cells.AddRange(newCells);
        }

        private Cell Get(int x, int y)
        {
            var existingCell = _cells.SingleOrDefault(c => c.X == x && c.Y == y);
            return existingCell ?? new Cell(x, y);
        }

        public IEnumerable<Cell> GetNeighbours(Cell cell)
        {
            yield return Get(cell.X - 1, cell.Y - 1);
            yield return Get(cell.X, cell.Y - 1);
            yield return Get(cell.X + 1, cell.Y - 1);
            yield return Get(cell.X + 1, cell.Y);
            yield return Get(cell.X + 1, cell.Y + 1);
            yield return Get(cell.X, cell.Y + 1);
            yield return Get(cell.X - 1, cell.Y + 1);
            yield return Get(cell.X - 1, cell.Y);
        }

        public void Clean()
        {
            var isolatedCells = _cells
                .Where(cell => GetNeighbours(cell).All(n => !n.IsAlive))
                .ToArray();

            _cells.RemoveAll(c => isolatedCells.Contains(c));
        }
    }
}

File: game.of.life/game.of.life.ts.web/scripts/GoL/RectangularInfinite2DGrid.ts
/// <reference path="../../typings/linq.d.ts"/>

'use strict';

namespace GoL {
	export class RectangularInfinite2DGrid {
		public Cells: Cell[] = [];

		constructor(cells: Cell[]) {
			this.Clean();
			for (let i = 0; i < cells.length; i++) {
				this.Cells.push(cells[i]);
			}
		}

		public Discover(): void {
			//var newCells = 
			//	(From cell in this.Cells
			//	From1 neighbour in this.GetNeighbours(cell)
			//	where!this.Cells.Contains(neighbour)
			//	select neighbour).Distinct().ToArray();
			
			const newCells: Cell[] = [];
			for (let i = 0; i < this.Cells.length; i++) {
				const cell = this.Cells[i];
				const neighbours = this.GetNeighbours(cell);
				for (let j = 0; j < neighbours.length; j++) {
					const neighbour = neighbours[j];

					for (let k = 0; k < this.Cells.length; k++) {
						const cell2 = this.Cells[k];

						if (!Cell.equals(neighbour, cell2)) {
							newCells.push(neighbour);
						}
					}
				}
			}

			const newDisctinctCells: Cell[] = this.uniqueCells(newCells);
			Enumerable.from(newDisctinctCells).forEach((cell: Cell) => this.Cells.push(cell));
		}

		private uniqueCells(cells: Cell[]) {
			let dict: any = {};

			for (let i = 0; i < cells.length; i++) {
				dict[i] = cells[i];
			}

			let newDisctinctCells: Cell[] = [];
			for (let key in dict) {
				newDisctinctCells.push(dict[key]);
			}

			return newDisctinctCells;
		}

		public GetNeighbours(cell: Cell): Cell[] {
			const neighbours: Cell[] = [];
			neighbours.push(this.Get(cell.X - 1, cell.Y - 1));
			neighbours.push(this.Get(cell.X, cell.Y - 1));
			neighbours.push(this.Get(cell.X + 1, cell.Y - 1));
			neighbours.push(this.Get(cell.X + 1, cell.Y));
			neighbours.push(this.Get(cell.X + 1, cell.Y + 1));
			neighbours.push(this.Get(cell.X, cell.Y + 1));
			neighbours.push(this.Get(cell.X - 1, cell.Y + 1));
			neighbours.push(this.Get(cell.X - 1, cell.Y));

			return neighbours;
		}

		public Get(x: number, y: number): Cell {
			const foundCell = Enumerable.from(this.Cells).singleOrDefault((c: Cell) => c.X === x && c.Y === y);
			return foundCell ? foundCell : new Cell(x, y);
		}

		public Clean(): void {
			const isolatedCells = Enumerable.from(this.Cells)
				.where((cell: Cell) => Enumerable.from(this.GetNeighbours(cell)).all((n: Cell) => !n.IsAlive()))
				.toArray();

			for (let i = 0; i < isolatedCells.length; i++) {
				const isolatedCell: Cell = isolatedCells[i];
				for (let j = 0; j < this.Cells.length; j++) {
					const cell = this.Cells[j];
					if (Cell.equals(isolatedCell, cell)) {
						this.Cells.splice(j, 1);
					}
				}
			}
			//Enumerable.from(this.Cells).removeAllRanges((c: Cell) => Enumerable.from(isolatedCells).contains(c));
		}

		//public override string ToString()
		//{
		//    return string.Join(", ", _cells.Where(c => c.IsAlive()));
		//}
	}
}
